core/storage/trie/polkadot_trie_db/polkadot_trie.hpp

```diff
@@ -0,0 +1,98 @@
+   */
+  class PolkadotTrie
+      : public face::ReadableMap<common::Buffer, common::Buffer>,
+        public face::WriteableMap<common::Buffer, common::Buffer> {
+    using NodePtr = std::shared_ptr<PolkadotNode>;
+    using BranchPtr = std::shared_ptr<BranchNode>;
+    using ChildRetrieveCallback =
+        std::function<outcome::result<NodePtr>(BranchPtr, uint8_t)>;
+
+    inline static outcome::result<NodePtr> DefaultChildRetrieveCallback(
+        const BranchPtr &parent, uint8_t idx) {
+      return parent->children.at(idx);
+    }
+
+   public:
+    enum class Error { INVALID_NODE_TYPE = 1 };
+
+   public:
+    /**
+     * Creates an empty Trie
+     * @param f a functor that will be used to obtain a child of a branch node
+     * by its index. Most useful if Trie grows too big to occupy main memory and
+     * is stored on an external storage
+     */
+    explicit PolkadotTrie(
+        ChildRetrieveCallback f = DefaultChildRetrieveCallback);
+
+    explicit PolkadotTrie(
+        NodePtr root, ChildRetrieveCallback f = DefaultChildRetrieveCallback);
+
+    NodePtr getRoot();
+
+    outcome::result<void> clearPrefix(const common::Buffer &prefix);
+
+    // value will be copied
+    outcome::result<void> put(const common::Buffer &key,
+                              const common::Buffer &value) override;
+
+    outcome::result<void> put(const common::Buffer &key,
+                              common::Buffer &&value) override;
+
+    outcome::result<void> remove(const common::Buffer &key) override;
+
+    outcome::result<common::Buffer> get(
+        const common::Buffer &key) const override;
+
+    bool contains(const common::Buffer &key) const override;
+
+   private:
+    outcome::result<NodePtr> insert(const NodePtr &parent,
+                                    const common::Buffer &key_nibbles,
+                                    NodePtr node);
+
+    outcome::result<NodePtr> updateBranch(BranchPtr parent,
+                                          const common::Buffer &key_nibbles,
+                                          const NodePtr &node);
+
+    outcome::result<NodePtr> deleteNode(NodePtr parent,
+                                        const common::Buffer &key_nibbles);
+    outcome::result<NodePtr> handleDeletion(const BranchPtr &parent,
+                                            NodePtr node,
+                                            const common::Buffer &key_nibbles);
+    // remove a node with its children
+    outcome::result<NodePtr> detachNode(const NodePtr &parent,
+                                        const common::Buffer &prefix_nibbles);
+    outcome::result<NodePtr> getNode(NodePtr parent,
+                                     const common::Buffer &key_nibbles) const;
+
+    uint32_t getCommonPrefixLength(const common::Buffer &pref1,
+                                   const common::Buffer &pref2) const;
+
+    NodePtr root_;
+    ChildRetrieveCallback retrieveChild;
```

Comment:
```suggestion
    ChildRetrieveCallback retrieve_child_;
```

Reply:
It's more of a function that a variable, so I think this naming suits it better. Especially if to look at its invocations.

Reply:
But still, it is a field of the class, so it should be named according to the rules

Reply:
I think I've found a compromise.